File: frame.c
#include <frame.h>
#include <measurement.h>

char schedule[15][24] = {
    {12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12},
    {12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12},
    { 8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8},
    { 8,  0,  8,  0,  8,  0,  8,  0,  8,  0,  8,  0,  8,  0,  8,  0,  8,  0,  8,  0,  8,  0,  8,  0},
    { 0,  8,  0,  8,  0,  8,  0,  8,  0,  8,  0,  8,  0,  8,  0,  8,  0,  8,  0,  8,  0,  8,  0,  8},
    {12, 12,  0,  0, 12, 12,  0,  0, 12, 12,  0,  0, 12, 12,  0,  0, 12, 12,  0,  0, 12, 12,  0,  0},
    { 0,  0, 12, 12,  0,  0, 12, 12,  0,  0, 12, 12,  0,  0, 12, 12,  0,  0, 12, 12,  0,  0, 12, 12},
    {12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0},
    {12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12},
    { 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0, 12, 12, 12, 12,  0,  0,  0,  0,  0,  0,  0,  0},
    {12, 12, 12, 12,  0,  0,  0,  0,  0,  0,  0,  0, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12},
    { 0,  0,  0,  0, 12, 12, 12, 12, 12,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0},
    { 0,  0,  0,  0,  0,  0,  0,  0,  0, 12, 12, 12,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0},
    { 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0, 12, 12, 12, 12,  0,  0,  0,  0},
    { 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0, 12, 12, 12, 12},
};

char frames[24] = {
    0x01, 0x10, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17, 0x18, 0x19, 0x1A, 0x1B, 0x1C, 0x1d, 0x1e, 0x1f, 0x20, 0x21, 0x22, 0x23, 0x24, 0x25, 0x26
};

void shift_left(unsigned char *ar, int size, int shift)
{
    while (shift--) {                           
        int carry = 0;                              
        for (int i = size - 1; i >= 0; --i) {   
            int next = (ar[i] & 0x80) ? 0x01 : 0;
            ar[i] = carry | (ar[i] << 1);
            carry = next;
        }   
    }
}

unsigned short take(struct Measurement* m, int i) {
    switch(i) {
        case 0:
            return m->Pm2_5;
        case 1:
            return m->Pm10;
        case 2:
            return m->Noise;
        case 3:
            return m->Solar;
        case 4:
            return m->Battery;
        case 5:
            return m->O3;
        case 6:
            return m->S02;
        case 7:
            return m->NO2;
        case 8:
            return m->CO;
        case 9:
            return m->Ammonia;
        case 10:
            return m->Temperature;
        case 11:
            return m->Humidity;
        case 12:
            return m->Barometric_Pressure;
        case 13:
            return m->VOX_CO2;
        case 14:
            return m->CO2;
        case 15:
            return -1;
        default:
            return -1;
    }
}

void push(unsigned char* buffer, int sizeOfBuffer, unsigned short value, int bits) {
    shift_left(buffer, sizeOfBuffer, bits);
    buffer[sizeOfBuffer - 1] = (unsigned char) value;
    if(bits == 12) {
        buffer[sizeOfBuffer - 2] = buffer[sizeOfBuffer - 2] | (value >> 8) & 0x0f;
    }
}

unsigned char* frame(unsigned char frameIndex, struct Measurement* measurement ) {
    static unsigned char frameToSend[12] = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
    for(int i = 0; i < sizeof(frameToSend); i++) {
        frameToSend[i] = 0;
    }
    push(frameToSend, sizeof(frameToSend), frames[frameIndex], 8);
    for(int i = 0; i < sizeof(schedule)/sizeof(frames); i++) {
        int bits = schedule[i][frameIndex];
        if(bits > 0) {
            unsigned short value = take(measurement, i);
            push(frameToSend, sizeof(frameToSend), value, bits);
        }
    }
    return frameToSend;
}

File: frame.js
'use strict';

const shl = (buf /*:Buffer*/, shiftBits /*:number*/) => {
    if (shiftBits < 0) { return shr(buf, -shiftBits); }
    if (shiftBits !== (shiftBits | 0)) { throw new Error("shiftBits must be a 32 bit int"); }
    const bytes = 2 * ((shiftBits >> 4) + ((shiftBits & 15) !== 0));
    const bits = (bytes * 8) - shiftBits;
    for (let reg = 0, i = 0; i - bytes < buf.length; i += 2) {
        reg <<= 16;
        if (i < buf.length - 1) {
            reg |= buf.readUInt16BE(i);
        } else if (i < buf.length) {
            reg |= buf[i] << 8;
        }
        if (i - bytes >= 0) {
            if (i - bytes < buf.length - 1) {
                buf.writeUInt16BE((reg >>> bits) & 0xffff, i - bytes);
            } else {
                if (i - bytes !== buf.length - 1) { throw new Error(); }
                buf[i - bytes] = reg >>> (bits + 8);
            }
        } else if (i - bytes === -1) {
            buf[0] = reg >>> bits;
        }
    }
};

const shr = (buf /*:Buffer*/, shiftBits /*:number*/) => {
    if (shiftBits < 0) { return shl(buf, -shiftBits); }
    if (shiftBits !== (shiftBits | 0)) { throw new Error("shiftBits must be a 32 bit int"); }
    const bytes = 2 * ((shiftBits >> 4) + ((shiftBits & 15) !== 0));
    const bits = 16 - ((bytes * 8) - shiftBits);
    for (let reg = 0, i = buf.length - 2; i + bytes >= -1; i -= 2) {
        reg >>>= 16;
        if (i >= 0) {
            reg |= buf.readUInt16BE(i) << 16;
        } else if (i === -1) {
            reg |= buf[0] << 16;
        }
        if (i + bytes + 1 < buf.length) {
            if (i + bytes >= 0) {
                buf.writeUInt16BE((reg >>> bits) & 0xffff, i + bytes);
            } else {
                if (i + bytes + 1) { throw new Error(); }
                buf[0] = reg >>> bits;
            }
        } else if (i + bytes + 1 === buf.length) {
            buf[i + bytes] = reg >>> (bits + 8);
        }
    }
};

const schedule = [
    [12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12 , 12, 12, 12, 12, 12, 12, 12, 12],
    [12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12],
    [ 8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8],
    [ 8,  0,  8,  0,  8,  0,  8,  0,  8,  0,  8,  0,  8,  0,  8,  0,  8,  0,  8,  0,  8,  0,  8,  0],
    [ 0,  8,  0,  8,  0,  8,  0,  8,  0,  8,  0,  8,  0,  8,  0,  8,  0,  8,  0,  8,  0,  8,  0,  8],
    [12, 12,  0,  0, 12, 12,  0,  0, 12, 12,  0,  0, 12, 12,  0,  0, 12, 12,  0,  0, 12, 12,  0,  0],
    [ 0,  0, 12, 12,  0,  0, 12, 12,  0,  0, 12, 12,  0,  0, 12, 12,  0,  0, 12, 12,  0,  0, 12, 12],
    [12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
    [12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12],
    [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0, 12, 12, 12, 12,  0,  0,  0,  0,  0,  0,  0,  0],
    [12, 12, 12, 12,  0,  0,  0,  0,  0,  0,  0,  0, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12],
    [ 0,  0,  0,  0, 12, 12, 12, 12, 12,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
    [ 0,  0,  0,  0,  0,  0,  0,  0,  0, 12, 12, 12,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
    [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0, 12, 12, 12, 12,  0,  0,  0,  0],
    [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0, 12, 12, 12, 12],
];

const frames = [
    0x01, 0x10, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17, 0x18, 0x19, 0x1A, 0x1B, 0x1C, 0x1d, 0x1e, 0x1f, 0x20, 0x21, 0x22, 0x23, 0x24, 0x25, 0x26
];

const names = ["Pm2_5", "Pm10", "Noise", "Solar", "Battery", "O3", "S02", "NO2", "CO", "Ammonia", "Temperature", "Humidity", "Barometric_Pressure", "VOX_CO2", "CO2"];

const parse = (data) => {
    const buffer = Buffer.from(data, 'hex');
    const frameId = buffer[0];
    shl(buffer, 8);

    const frameIndex = frames.indexOf(frameId);
    let temp = {};
    for(let measurementIndex = 0; measurementIndex < schedule.length; measurementIndex++) {
        const bits = schedule[measurementIndex][frameIndex];
        const name = names[measurementIndex];
        if(bits == 8) {
            temp[name] = format(name, buffer[0]);
        } else if(bits == 12) {
            temp[name] = format(name, 16 * buffer[0] + (0xf0 & buffer[1]) / 16);
        }
        shl(buffer, bits);
    }
    return temp;
}

const metrics = {
    "Pm2_5": {offset: 0, step: 1, unit:"ug/m3"},
    "Pm10": {offset: 0, step: 1, unit:"ug/m3"},
    "Noise": {offset: 0, step: 0.5, unit:"dB"},
    "Solar": {offset: 0, step: 0.5, unit:"%"},
    "Battery": {offset: 0, step: 0.5, unit:"%"},
    "O3": {offset: 0, step: 5, unit:"ppb"},
    "S02": {offset: 0, step: 10, unit:"ppb"},
    "NO2": {offset: 0, step: 5, unit:"ppb"},
    "CO": {offset: 0, step: 0.5, unit:"ppm"},
    "Ammonia": {offset: 0, step: 0.5, unit:"ppm"},
    "Temperature": {offset: -100, step: 0.2, unit:"oC"},
    "Humidity": {offset: 0, step: 0.2, unit:"RH"},
    "Barometric_Pressure": {offset: 600, step: 0.5, unit:"hPA"},
    "VOX_CO2": {offset: 0, step: 1, unit:"ppm"},
    "CO2": {offset: 0, step: 0.5, unit:"ppm"}
  };

const format = (metric_id, value) => {
    const result = ((value + metrics[metric_id].offset) * metrics[metric_id].step).toFixed(2);
    return result;
}

const result = parse("16012013314a01d0070041e0");
console.log(JSON.stringify(result));
